book/src/cleaning.md

```diff
@@ -0,0 +1,62 @@
+# TODO Cleaning up Resource Leaks
+
+Since Java doesn’t automatically free the memory on Rust’s heap when the
+`Counter` garbage collected, there could be a memory leak. Many native resources
+need similar treatment, and so we are going to automate it using the [Cleaner
```

Comment:
Even if we don't cover the approach with explicit memory management, I think it shall be mentioned with a todo to expand on this topic, with some considerations on choosing the right approach.

Cleaner, as any GC-based approach, is not universally applicable:
- Cleaning actions are performed in a separate thread, hence the Java class has to be thread-safe
- If the app creates Java proxies of dependent native objects, they likely have to be destroyed in the reverse order of their creation — something that is possible with PhantomReferenceQueues, but will add quite some complexity
- Any finalization-based approach is not suitable for high-volume de-allocation rate. I no longer have the benchmarks, as they are in private Exonum repository, but about three years ago we also considered which approach to take o simplify the memory management, and prototyped:
    - Explicit (it may be eventually hidden in a single place if all native peers are created in a certain bounded context — that's how EJB works)
    - GC-based approaches:
      - `finalize`
      - `PhantomReferenceQueue`, no explicit clean action
      - `PhantomReferenceQueue`, with explicit clean action

Any GC-based approaches were at least an order of magnitude slower. The GC-based approaches compared in the following way, from fastest to slowest:
1. `PhantomReferenceQueue`, with explicit clean action
2. `finalize`
3. `PhantomReferenceQueue`, no explicit clean action
      
See also [my answer](https://stackoverflow.com/a/44799935) from the time we researched this topic in Exonum.

Reply:
Also, we can consider finishing this chapter in subsequent PRs, so that it doesn't grow too large.

Reply:
Yeah, we'll need to expand on this over time. I'll edit it to state that the chapter isn't a complete guide to memory management of native resources, and describe the upsides/downsides of the approach, then mention the alternatives you've described. Then, we can expand on those over time - potentially in chapters that revisit the same topic, or by additions to this particular chapter.